### ISSUE TITLE
Change reports for multiline to be able to output potential fails/pass

### ISSUE THREAD

**Issue by TDiazT** [TDiazT/CoqR-Tester#48]
As it is right now, expressions ran by multiline strategy don't get the appropriate treatment with potential fails and passes. That's because reports generated have a format:
```python
{
    "expression": "e1",
    "sub_reports": {
        "expression": "e1",
        ...
    },
    ...
}

```

Then, when the sub reports are being compared, there is no potential anything, because there is always just one expression...

**Comment by TDiazT** [comment]
@Mbodin  Ok, so here's what there is.

In a line by line case, you usually have small sequences of expressions, such as `e1;e2;e3`. This allows to store the context because there's not much data to store. For example, for `e3` you would get the context `e1;e2` (actually I have to fix this haha).

The problem with the multiline expressions is that you get expressions that span several lines and in some cases, you can have a big file. In those cases, the last expression would consider its context to be the whole file... :confused: 

I'm thinking that maybe I can include all this information during the whole process but after the comparison I would clean this up. In this case I could probably include a `strategy` field to the reports and in case it was multiline, delete the context for every expression.

This would mean that when checking the results, you wouldn't be able to see the context for an expression with multiline (otherwise ok for the line by line strategy) but still preserve the `potentials`.

What do you think? Is it worth differentiating the potential passes/fails from the actual ones?

**Comment by Mbodin** [comment]
OK, there are two different things here.

On one hand, the differentiation between fails and passes is useful to evaluate our formalisation. To get the full grade, we would like no fail to appear (but potential fails are fine). In some sense, one fail could discredit the trust one can have on our framework. Of course, we can still argue about individual fails, but we would like to have as little as possible.

On the other hand, we would like to have a useful testing framework. Fails are interesting as they inform us where to focus, and this helps development. Potential fails are useful too, but less. Expression context can be very useful: it avoids us to actually look in the file to understand what is going on. We can also add the column, to help differentiating two similar expressions (as in `<something>; a; <something>; a`: an error on `a` provides us very little information).

So I think that we have to choose some helpful information to show to the user, but if possible, try to restrict what we call “fail” to the ones we are sure are fails. So my intuition would be to move any suspicious fails into the “potential fail” category, but try making this category easier to use. For instance, adding the context of the current line in multiline mode would help. If this is possible, we could also add the environment (which variables are defined, and what are their values), or something similar. It is difficult to know what would be useful in general, but we can try adding the information we can (if it is not too big/difficult to get). A compromise could be to add as a context in multiline mode the last 3–4 executed expressions and their results. Would it be easy to add in the current framework?

**Comment by TDiazT** [comment]
I think there might be an issue in multiline, where you could have an error at the beginning of the file and then everything else could be tagged as "potential". How can you tell that the rest is certainly a fail or a pass?

I agree with what you're saying but it is not clear to me how to go for it.

**Comment by TDiazT** [comment]
Maybe that last idea could work...

We could say that a potential fail/pass is something that fails/passes considering there was a failure in one of the last 3-4 expressions. Although it could still happen that:
1. A potential pass is really a pass and has no relation to the expression that failed 
1. A pass was actually a potential pass, but the failure came from before the last 4 expressions

**Comment by Mbodin** [comment]
I agree that this is a frustrating solution. The correct solution would probably be to mark the whole file as “Potential”… but this is frustrating. :-\

So here is a question: is it frequent that there is an error in the middle of a large mutliline file?

**Comment by TDiazT** [comment]
Based on a quick look, yeah, it is frequent at the middle or beginning. 

At least in the R test files.
